src/bl_install.c

```diff
@@ -238,6 +249,107 @@ static int bootloader_send(FILE *fp, int length, u32* pCrc32)
+	if (read_bytes != FLASH_SIZE) {
+		printf("WARNING: firmware file might be truncated: %d bytes expected, %d bytes read\n", FLASH_SIZE, read_bytes);
+	}
+
+	int err = ERR_UNK;
+	if (verbose) {
+		// Sector-by-sector CRC comparison.
+		// Compares the remote and local CRCs of each sector, showing the results of each.
+		printf("Sector CRC comparison:\n");
+		bool all_ok = true;
+		const char *last_area = NULL;
+
+		for (u32 sector = starting_sector; sector < starting_sector + num_sectors; sector++) {
+			u32 local_sector_crc32 = crc32(0, firmware + sector * FLASH_SECTOR_SIZE, FLASH_SECTOR_SIZE);
+			u32 remote_sector_crc32 = 0;
+			int sector_err = bootloader_checksum(sector * FLASH_SECTOR_SIZE, FLASH_SECTOR_SIZE, &remote_sector_crc32);
+
+			bool this_ok = remote_sector_crc32 == local_sector_crc32;
+			all_ok = all_ok && this_ok;
+
+			const char *area = sector_explanation(sector);
+			if (area != last_area) { // static string literals can be compared by pointers
+				printf("=== %s%s ===\n", area, sector_is_mutable(sector) ? " (CRC mismatch is expected here)" : "");
+				last_area = area;
+			}
+
+			if (sector_err == ERR_UNK) {
+				printf("Sector %3d: Remote CRC = %08X, local CRC = %08X%s\n", sector, remote_sector_crc32,
+					local_sector_crc32, this_ok ? "" : " ERR");
+			} else {
+				printf("Error requesting sector CRC.  Err = %i\n", sector_err);
+				err = sector_err;
+			}
+		}
+
+		if (all_ok) {
+			puts("All sectors match.");
+		} else {
+			puts("Some sectors do not match.");
+		}
+	} else {
+		// Do a single comparison, calculating the CRC over multiple sectors.
+		u32 local_crc32 = crc32(0, firmware + starting_sector * FLASH_SECTOR_SIZE, num_sectors * FLASH_SECTOR_SIZE);
+		printf("Requesting remote CRC...\n");
+		u32 remote_crc32 = 0;
+		err = bootloader_checksum(starting_sector * FLASH_SECTOR_SIZE, num_sectors * FLASH_SECTOR_SIZE, &remote_crc32);
+		if (err == ERR_UNK) {
+			printf("Remote CRC = %08X, local CRC = %08X%s\n", remote_crc32, local_crc32,
+				remote_crc32 == local_crc32 ? "" : " ERR");
+		} else {
+			printf("Error requesting full CRC.  Err = %i\n", err);
+		}
+	}
+
+	free(firmware);
+	return err;
+}
+
+static const char *sector_explanation(u32 sector_number)
+{
+	const char *explanation = "unknown";
+	if (sector_number < 0x04) { // 0x000000 - 0x03FFFF
+		explanation = "U-Boot";
+	} else if (sector_number == 0x04) { // 0x040000 - 0x04FFFF
+		explanation = "U-Boot environment";
+	} else if (sector_number < 0x25) { // 0x050000 - 0x24FFFF
+		explanation = "Linux kernel";
+	} else if (sector_number < 0xCB) { // 0x250000 - 0xCAFFFF
+		explanation = "Root file system";
+	} else if (sector_number < 0xFA) { // 0xCB0000 - 0xF9FFFF
+		explanation = "User data";
+	}
+	return explanation;
+}
+
+static bool sector_is_mutable(u32 sector_number)
+{
+	bool mutable = false;
+	if (sector_number == 4) {
+		// U-Boot environment
+		mutable = true;
+	} else if (sector_number >= 0xCB) {
+		// User data
+		mutable = true;
+	}
```

Comment:
Since the sectors aren't fixed, it would make more sense to look up these by name or have a sector info struct that contains both the name and a mutable bool flag.

Reply:
Hmm, agreed; I was initially thinking that the table would be an unnecessary complication of the code, but it may be a bit cleaner after all.